File: Texture.hs
module Texture where

import Codec.Picture

import Vec3

data Texture = ConstantTexture Vec3 
             | CheckeredTexture Texture Texture 
             | ImageTexture (Image PixelRGB8)

textureValue :: Texture -> Vec3 -> Vec2 -> Vec3

textureValue (ConstantTexture c) _ _ = c 

textureValue (CheckeredTexture t0 t1) (px, py, pz) hitTexCoord = 
    let sines = (sin (10 * px)) * (sin (10 * py)) * (sin (10 * pz)) 
    in if sines < 0 then
           textureValue t0 (px, py, pz) hitTexCoord
       else
           textureValue t1 (px, py, pz) hitTexCoord

textureValue (ImageTexture img) _ (u, v) = 
    let width = imageWidth img
        height = imageHeight img
        i = round ((max (min u 1.0) 0.0) * ((fromIntegral width) - 1.0))
        j = round ((max (min v 1.0) 0.0) * ((fromIntegral height) - 1.0))
        (PixelRGB8 pr pg pb) = pixelAt img i j
    in ((fromIntegral pr) / 255.0, (fromIntegral pg) / 255.0, (fromIntegral pb) / 255.0)

File: raytracer-cpp/src/texture.cpp
#include "texture.h"

Vec3 ConstantTexture::value(const Vec2 &t, const Vec3 &p) {
    return color;   
}

Vec3 CheckeredTexture::value(const Vec2 &t, const Vec3 &p) {
    float sines = sin(10 * p.x) * sin(10 * p.y) * sin(10 * p.z); 

    if (sines < 0) {
        return texture0->value(t, p);
    }
    else {
        return texture1->value(t, p);
    }
}

Vec3 ImageTexture::value(const Vec2 &t, const Vec3 &p) {
    int x = width * t.x;
    int y = height * t.y;

    unsigned char r = data[3 * (x * height + y) + 0]; 
    unsigned char g = data[3 * (x * height + y) + 1]; 
    unsigned char b = data[3 * (x * height + y) + 2]; 

    return Vec3(r / 255.0, g / 255.0, b / 255.0);
}
